Feat: Filter groups

w33ble opened the pull request:
Closes #23117

~~Blocked by https://github.com/elastic/kibana/pull/34823~~ merged

Allow grouping of filters and group selection in the `filters` function.

![Apr-18-2019 09-49-33-optimized](https://user-images.githubusercontent.com/404731/56377654-aa943c00-61bf-11e9-9c40-7570393c1326.gif)

- The `filters` function now accepts an argument, which is multi and has an unnamed alias. Use it to select which group(s) of filters you want to use, like `filters group1 | demodata`
- The time picker and dropdown functions both take a `filterGroup` argument, which is passed along to the filter expression

---

## Testing

Grouping works with simple string matching (case sensitive). If you specify a group (or many groups) on the `filters` function, only filters that belong to the group(s) will be applied. If no group is provided, it behaves like global filters do now.

The easiest way is to simply watch filter values update. Create 3 elements with the following expressions:

1. `filters`
1. `filters group1`
1. `filters group2`

Now add a time picker or dropdown element (or both, or multiples of one or both... whatever you want). You will see filters applied to `#1`, but not `#2` or `#3`.

Now add a filter group to one (or several) or the time pickers and/or dropdowns. You can **use the sidebar**, as in the gif above, and you should see both "group1" and "group2" in the dropdown. If you add a different `filterGroup` to any of the filter elements, that will show up in the dropdown as well.

You can also **add filterGroups to the expression**, by adding `filterGroup=group1` to the `timefilterControl`/`dropdownControl` function, like in this example:

```
timefilterControl compact=true column=@timestamp filterGroup=group1 | render
```

Now that should show up in `#1` (since that one is global) and `#2`. Change the value to `group2` and it should show up in `#1` and `#3`. Change it to some other value and it will again only show up in `#1`.

The value for `filterGroup` can be *any* string, and if you create other groups (ie. use a string other than "group1" and "group2"), they should show up in the dropdown on the sidebar as well.

To check using multiple groups, set at least one filter on each of `group1` and `group2`. Now update `#3` to this expression: `filters group2 group1`. `#3` should now show filters in both `group1` and `group2`, and match `#1` (unless you've added more groups).

Be sure to try using `filters` with multiple groups, and adding multiple filters to a group.

### Sidebar

The sidebar control for the groups will show a dropdown of all groups that exist on any filters, as well as any groups being consumed by any `filters` functions (even if they are nested), which makes it easy to pick existing groups from the sidebar. Users can also create new groups on the fly from the sidebar. 

If the element already uses a filter, and it's the only one, and you change it, the old group value will no longer show up in the list. You will have to manually add it back. This is also demonstrated in the gif above.

w33ble committed:
feat: add group arg to filters function

multi-arg so multiple groups can be used, and the value is always an array

test: add setFilter tests

w33ble committed:
chore: replace getGlobalFilterExpression with getGlobalFilters

w33ble committed:
feat: add group arg to filters function

w33ble commented on x-pack/plugins/canvas/canvas_plugin_src/renderers/dropdown_filter/index.tsx:
```diff
@@ -29,11 +30,18 @@ export const dropdownFilter: RendererFactory<Config> = () => ({
+      const filterAST = fromExpression(filter);
       value = get(filterAST, 'chain[0].arguments.value[0]');
+
+      const filterGroup = get(filterAST, 'chain[0].arguments.filterGroup[0]');
```
This is unfortunately becoming a common pattern, where expression args need to be synced with the filter expression at render time. Not sure how to abstract it though ☹️

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9090/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9111/)

w33ble committed:
feat: add grouping to filter functions

timefilter, exacttly, dropdownControl, and timefilterControl

w33ble committed:
chore: add resolved config filter sync helper

w33ble committed:
feat: add filterGroup to filter expressions

and keep the values in sync with expression via helper function

w33ble commented on x-pack/plugins/canvas/canvas_plugin_src/renderers/dropdown_filter/index.tsx:
```diff
@@ -20,20 +21,40 @@ interface Config {
+    if (filterExpression !== '') {
+      // NOTE: setFilter() will cause a data refresh, avoid calling unless required
+      // compare expression and filter, update filter if needed
+      const { changed, newAst } = syncFilterExpression(config, filterExpression, [
```
This is unfortunately becoming a common pattern, where expression args need to be synced with the filter expression at render time. I made a helper but I'm not sure how to abstract it better ☹️

w33ble commented on x-pack/plugins/canvas/canvas_plugin_src/renderers/dropdown_filter/index.tsx:
```diff
@@ -20,20 +21,37 @@ interface Config {
+    if (filterExpression !== '') {
+      // NOTE: setFilter() will cause a data refresh, avoid calling unless required
+      // compare expression and filter, update filter if needed
+      const { changed, newAst } = syncFilterExpression(config, filterExpression, ['filterGroup']);
```
This is unfortunately becoming a common pattern, where expression args need to be synced with the filter expression at render time. I made a helper but I'm not sure how to abstract it better ☹️

clintandrewhall replied:
Might consider creating a `discuss` issue and referencing it here and elsewhere...?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9126/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9129/)

w33ble committed:
feat: add basic group input to sidebar

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9894/)

w33ble committed:
feat: add getGlobalFilterGroups selector, with tests

w33ble committed:
chore: remove unused argForm resolve values

w33ble committed:
feat: pass filter groups into argTypes

added via FunctionForm, which renders the associated argType

w33ble committed:
feat: add filterGroup arg type

use it for the timepicker and dropdown UIs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10279/)

elasticmachine commented:
Pinging @elastic/kibana-canvas

ryankeairns commented:
@w33ble Can you provide some steps on how to use this? It's not clear to me how you name a group in order for it to display in the dropdown, and I'm also not sure what 'Create a group' does. Thanks for clarifying.

One quick note, do you think this would be used frequently enough that the input could just be displayed all the time as opposed to being behind the plus icon?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10302/)

w33ble commented:
> One quick note, do you think this would be used frequently enough that the input could just be displayed all the time as opposed to being behind the plus icon?

@ryankeairns hard to say. Seems like something that would get a bunch of use, but it's hard to tell. Having it show up by default would be easy though, just need to add it to the element expressions.

I'll add more details to the description, sorry about that.

w33ble commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10307/)

ryankeairns reviewed (changes requested):
This is a nice addition, I mostly tripped over the terminology and it was unknown how exactly to name filters. 

After getting it working, I felt as though it would be more straightforward to just give the filter a name or ID in the side panel and type that into each expression. Then, I realized that is essentially what is happening here and that the select input is just providing you a means to use existing group names... that sounds obvious now, but I can assure you it was a mindbender :)

Suffice it all to say, I think some clarifications in the label, placeholder, and help text would have saved me some deep thoughts. Here is a mockup capturing my recommendations:

<img width="611" alt="Screenshot 2019-04-18 16 12 04" src="https://user-images.githubusercontent.com/446285/56391624-d0880380-61f4-11e9-821a-c073401f0861.png">

- Adding the word 'name' may seem trivial, but that implies that I must name the filter in some way which is a useful hint. 
- Adding to the placeholder text (apply globally) further informs what is currently happening and what a selection here will thus change. 
- An example in the help text (`filters groupName` - you could wrap that in EuiCode) shows me exactly what I need to do.
- Lastly, moving the 'Create new' to the upper right keeps things better organized, visually, and is documented in EUI... i.e., it's a prop you can pass.

Let me know how that works!

ryankeairns commented:
Ah, I missed one other important observation... I added names to my filters, but they never appeared in the select dropdown. If I typed in a new group name via 'Create new group', one matching my expression, then the functionality worked correctly. The dropdown just never populated for me.

w33ble commented:
@ryankeairns thanks for the feedback, it's super insightful.

I can't move the create link there, and the help text lacks formatting :(. But what do you think of this?

![Screenshot 2019-04-18 16 40 05](https://user-images.githubusercontent.com/404731/56397487-cb2cb800-61f8-11e9-8827-b321def115af.png)

![Screenshot 2019-04-18 16 40 12](https://user-images.githubusercontent.com/404731/56397541-1050ea00-61f9-11e9-8f9f-13018413b960.png)

Still not crazy about the help text, but I can't think of a clearer way to say that... I think I'm too close to the feature. Input is appreciated. 

Ar for the dropdown vaoues, it doesn't work like you think, but I think I can make it work that way. I'll try and update the PR.

w33ble committed:
fix: tweak filter group name and help text

w33ble committed:
fix: remove aliases from filter functions

it breaks the filter lookup code, and the value is almost always set from the UI control anyway

w33ble committed:
fix: sort global filter groups

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10361/)

ryankeairns commented:
Wrapping up work on another PR, will get to this in the afternoon!

w33ble committed:
feat: include filters function groups

when collecting filter groups, also include groups in expressions

ryankeairns commented:
@w33ble  I'm still thinking through the text, but one quick question regarding how this is intended to work.

If I have two elements
1) with group name of **group1** 
2) with no group name

... and I select **group1** for the Filter group name, then shouldn't the filter ONLY apply to the element that has `filters group1'. In other words, it should no longer apply globally to all elements, right? 

I would expect that if I set a group name on the filter, then it will ONLY apply to elements with `filters group1`, but I might be misunderstanding this.

In this screenshot, I'm expecting the 'no group' metric to continue showing the cumulative total of 3,000: 

<img width="680" alt="Screenshot 2019-04-19 14 47 02" src="https://user-images.githubusercontent.com/446285/56441374-78173b80-62b2-11e9-87c3-94b0ef2c1407.png">

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10454/)

w33ble commented:
> In other words, it should no longer apply globally to all elements, right?

That's not right. If you use `filters` but don't apply any groups, you get *all* filters, whether they belong to a group or not. The `group` just acts as a way to filter what filters are applied.

That said, I could add a new arg that would ignore any filters that belong to any group. Then you'd have something like this:

- `filters` would apply *all* filters, both grouped and not
- `filters group1` would only apply filters that belong to `"group1"
- `filters ignoreGrouped=true` would only apply filters that have no group (this is what you were expecting)

It complicates things a little, but gives users more control without changing the existing functionality. Does that sound good to you?

cqliu1 reviewed (commented):
Just came across one bug. When I try to add a new filter group from a `dropdown_filter` and hit `set`, it reverts back to the dropdown input but still shows 'No group' as the value while the expression has been updated with the correctly new group name. This bug effects `dropdown_filter` and `advanced_filter`. `time_filter` seems to be working correctly.

![Apr-19-2019 13-03-50](https://user-images.githubusercontent.com/1697105/56442318-24522580-62a5-11e9-87e7-20a9ca532bf8.gif)

Another unrelated bug is that when you're typing in the `advanced_filter` and hit backspace, it deletes the element. This bug is actually fixed as part of #34140, so no action needed on your part.

ryankeairns commented:
Yeah, I think that would help, thanks for tinkering on it. 

I wrote a really long response where `filters` only includes those without a group name, but I think it's essentially a wash (both have +/-'s). Let's roll with your suggestion!

ryankeairns commented:
> > In other words, it should no longer apply globally to all elements, right?
> 
> That's not right. If you use `filters` but don't apply any groups, you get _all_ filters, whether they belong to a group or not. The `group` just acts as a way to filter what filters are applied.
> 
> That said, I could add a new arg that would ignore any filters that belong to any group. Then you'd have something like this:
> 
> * `filters` would apply _all_ filters, both grouped and not
> * `filters group1` would only apply filters that belong to `"group1"
> * `filters ignoreGrouped=true` would only apply filters that have no group (this is what you were expecting)
> 
> It complicates things a little, but gives users more control without changing the existing functionality. Does that sound good to you?

ryankeairns committed:
mostly copy changes

ryankeairns commented:
@w33ble Here is a PR with some minor copy and design edits:

https://github.com/w33ble/kibana/pull/8

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10469/)

w33ble committed:
feat: add ungrouped arg to filters

w33ble commented:
Added an `ungrouped` arg to filters, which when truthy, skips filters that have a group.

![Screenshot 2019-04-19 15 22 54](https://user-images.githubusercontent.com/404731/56446341-0392cb80-62b7-11e9-9b08-fe156d5a128b.png)

![Apr-19-2019 15-12-29-optimized](https://user-images.githubusercontent.com/404731/56446368-1c9b7c80-62b7-11e9-94c8-a9cbfbf43e6c.gif)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10491/)

w33ble commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10497/)

ryankeairns reviewed (approved):
👍 LGTM. Works as described, very useful.

ryankeairns commented:
@w33ble  I see the new `ungrouped` arg in the autocomplete, which is awesome, and that has me wondering how it gets documented in the external function docs: https://www.elastic.co/guide/en/kibana/current/canvas-client-functions.html#_filters

Likewise, we'd probably want to make mention of the `name` arg as well (e.g. `group1`).

I suspect that we just need to create a separate issue for docs?

cc: @shaunmcgough

w33ble commented:
> that has me wondering how it gets documented in the external function docs

That used to be automated, no idea how that works anymore though.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10900/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11936/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11944/)

clintandrewhall commented on x-pack/plugins/canvas/canvas_plugin_src/renderers/dropdown_filter/index.tsx:
```diff
@@ -20,20 +21,37 @@ interface Config {
+  }
+
+  const filterAST = fromExpression(filterExpression);
+  return get(filterAST, 'chain[0].arguments.value[0]', MATCH_ALL) as string;
```
This seems really brittle... `get` has its usefulness, but this *appears* to have a number of runtime error vulnerabilities.

clintandrewhall commented on x-pack/plugins/canvas/canvas_plugin_src/uis/arguments/filter_group.js:
```diff
@@ -0,0 +1,99 @@
+/*
```
New files without Typescript make me sad.

clintandrewhall commented on x-pack/plugins/canvas/public/lib/sync_filter_expression.ts:
```diff
@@ -0,0 +1,37 @@
+/*
```
Files with TS make me happy!

clintandrewhall commented on x-pack/plugins/canvas/public/lib/sync_filter_expression.ts:
```diff
@@ -0,0 +1,37 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// @ts-ignore
```
nit: add a comment to the right explaining the ignore, so we can find/fix later.

clintandrewhall commented on x-pack/plugins/canvas/public/lib/sync_filter_expression.ts:
```diff
@@ -0,0 +1,37 @@
+// @ts-ignore
+import { fromExpression } from '@kbn/interpreter/common';
+// @ts-ignore
+import { set, del } from 'object-path-immutable';
```
Does this not have a `@types` lib?  This is where the comment comes in handy.

clintandrewhall commented on x-pack/plugins/canvas/public/lib/sync_filter_expression.ts:
```diff
@@ -0,0 +1,37 @@
+import { get } from 'lodash';
+
+export function syncFilterExpression(
+  config: Record<string, any>,
```
Is this truly `any`?  Or is it `string | number | null`?

w33ble replied:
Yeah, pretty much. Nothing is used of enforced in this file, it just uses whatever values existing in the object. It can also be more than just `string | number | null`, at a minimum it's `string | number | null | Array<string | number | null>`, no idea what other kinds of values are valid there.

clintandrewhall reviewed (commented):
Great work.  Sorry it took me a bit to get to my review.  Consider my comments, but feel free to follow-up on larger items.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14128/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14136/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14150/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14168/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14172/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14177/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14179/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14181/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14183/)

w33ble commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14188/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14375/)

w33ble commented:
retest

w33ble committed:
chore: types and translations

w33ble committed:
chore: address nit feedback

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14423/)

monfera commented:
The referred feature request says #33605 is a blocker (open atm), is it still true?

w33ble commented:
@monfera good question. The other 2 blocker issues are fixed, but that one remains. https://github.com/elastic/kibana/issues/33605 still needs to get fixed, but it doesn't really stop this feature from working, so it's not actually a "blocker". 

The filter value also still doesn't reliably update all the time, which may or may not be related to that issue. I spent some time in the past trying to figure out why it happens but wasn't successful, sadly.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14737/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14748/)

w33ble commented:
Thanks to @cqliu1, I figured what new group values sometimes don't show up in the sidebar. I opened https://github.com/elastic/kibana/issues/36950 to track it.
